Unity-Technologies/ScriptableRenderPipeline
# OcclusionMesh for SRP VR + HDRP XRGraphicsConfig hookup fix

### Purpose of this PR
The occlusion mesh is a performance feature for VR which enables shaders to return early for pixels that invisible in the HMD being used. This changeset adds calls to ScriptableRenderContext.RenderOcclusionMesh to HDRP's depth prepass and to LWRP's depth prepass and forward opaque stages. This call correctly applies occlusion mesh scaling as configured in the SRP asset UI.
![image](https://user-images.githubusercontent.com/3450690/45189993-6b951580-b1f0-11e8-8ef7-30772896cb3f.png)

---
### Release Notes
This changeset depends on the following Editor changeset: https://ono.unity3d.com/unity/unity/changelog?branch=srp/renderocclusionmesh

We have to wait for the associated PR to complete before we can merge this changeset in:
https://ono.unity3d.com/unity/unity/pull-request/73569/_/srp/renderocclusionmesh

Furthermore we should consider whether a backport to 2018.3 or 2018.2 is necessary before merging this into SRP master. (Or should I bracket it with IF_2018_3_OR_NEWER?)

---
### Testing status
**Katana Tests**: First off we need to make sure the Katana SRP tests are green?
On side branch: https://katana.bf.unity3d.com/projects/com.unity.render-pipelines/builders?ScriptableRenderLoop_branch=occlusionmesh&unity_branch=srp%2Frenderocclusionmesh&automation-tools_branch=add-platform-filter

PR merged into trunk: https://katana.bf.unity3d.com/projects/com.unity.render-pipelines/builders?ScriptableRenderLoop_branch=occlusionmesh&unity_branch=trunk&automation-tools_branch=add-platform-filter

**Manual Tests**: What did you do?
Visually confirmed occlusion mesh in HRDP. There are artifacts in the occlusion mesh, but this is due to an existing bug in HDRP that is unrelated to the occlusion mesh call. Visually confirmed that occlusion mesh scaling was correctly applied for scaling values >= 1.
![image](https://user-images.githubusercontent.com/3450690/45190643-d4ca5800-b1f3-11e8-9297-5cda0ba090e6.png)

Same for LWRP, plus I toggled the "render occlusion mesh" setting in XRGraphicsConfig to confirm that the setting was being applied correctly. Visually confirmed that occlusion mesh scaling was correctly applied for scaling values >= 1.

**Automated Tests**: What did you setup?
None.

Any test projects to go with this to help reviewers?
Should be reproducible on an empty project. 

---
### Overall Product Risks
**Technical Risk**: 
Low: Adds an additional render call prior to certain render passes, but doesn't alter existing renders otherwise

**Halo Effect**: None, Low, Medium, High?
Low for same reason as above

---
### Comments to reviewers
The HRDP XRGraphicsConfig hookup is tied to this PR because occlusion mesh is the first configurable XR feature supported by HDRP, and so it only became possible to confirm whether the hookup was working after adding occlusion mesh.

Please give feedback on whether backports to previous unity versions are necessary for this PR.
